moveit_ros/occupancy_map_monitor/src/occupancy_map_monitor.cpp

```diff
@@ -34,129 +34,104 @@
 
 /* Author: Ioan Sucan, Jon Binney */
 
-#include <rclcpp/rclcpp.hpp>
-#include <moveit_msgs/srv/save_map.hpp>
-#include <moveit_msgs/srv/load_map.hpp>
 #include <moveit/occupancy_map_monitor/occupancy_map.h>
 #include <moveit/occupancy_map_monitor/occupancy_map_monitor.h>
+#include <moveit/occupancy_map_monitor/occupancy_map_monitor_middleware_handle.hpp>
+#include <moveit_msgs/srv/load_map.hpp>
+#include <moveit_msgs/srv/save_map.hpp>
+
 #include <boost/bind.hpp>
+#include <rclcpp/rclcpp.hpp>
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
 
 namespace occupancy_map_monitor
 {
+namespace
+{
 static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit.ros.occupancy_map_monitor");
+}
 
 OccupancyMapMonitor::OccupancyMapMonitor(const rclcpp::Node::SharedPtr& node, double map_resolution)
-  : map_resolution_(map_resolution), debug_info_(false), mesh_handle_count_(0), node_(node), active_(false)
+  : OccupancyMapMonitor{ std::make_unique<OccupancyMapMonitorMiddlewareHandle>(node, map_resolution, ""), nullptr }
 {
-  initialize();
 }
 
 OccupancyMapMonitor::OccupancyMapMonitor(const rclcpp::Node::SharedPtr& node,
                                          const std::shared_ptr<tf2_ros::Buffer>& tf_buffer,
                                          const std::string& map_frame, double map_resolution)
-  : tf_buffer_(tf_buffer)
-  , map_frame_(map_frame)
-  , map_resolution_(map_resolution)
-  , debug_info_(false)
-  , mesh_handle_count_(0)
-  , node_(node)
+  : OccupancyMapMonitor{ std::make_unique<OccupancyMapMonitorMiddlewareHandle>(node, map_resolution, map_frame),
+                         tf_buffer }
 {
-  initialize();
 }
 
-void OccupancyMapMonitor::initialize()
+OccupancyMapMonitor::OccupancyMapMonitor(std::unique_ptr<MiddlewareHandle> middleware_handle,
+                                         const std::shared_ptr<tf2_ros::Buffer>& tf_buffer)
+  : middleware_handle_{ std::move(middleware_handle) }
+  , tf_buffer_{ tf_buffer }
+  , map_frame_{ "" }
+  , map_resolution_{ 0.0 }
+  , debug_info_{ false }
+  , mesh_handle_count_{ 0 }
+  , active_{ false }
 {
-  /* load params from param server */
-  if (map_resolution_ <= std::numeric_limits<double>::epsilon())
-    if (!node_->get_parameter("octomap_resolution", map_resolution_))
-    {
-      map_resolution_ = 0.1;
-      RCLCPP_WARN(LOGGER, "Resolution not specified for Octomap. Assuming resolution = %g instead", map_resolution_);
-    }
-  RCLCPP_DEBUG(LOGGER, "Using resolution = %lf m for building octomap", map_resolution_);
+  if (middleware_handle_ == nullptr)
+  {
+    throw std::invalid_argument("OccupancyMapMonitor cannot be constructed with nullptr MiddlewareHandle");
+  }
+
+  // Get the parameters
+  map_frame_ = middleware_handle_->getMapFrameParameter();
```

Comment:
Things like this make me think that classes should just use structs for their parameter configs. The middleware handle could simply initialize those from `rclcpp::Parameters` or manually in case of unit tests.

Reply:
done, lmk what you think of the implementation